azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -180,41 +194,49 @@ def _on_mqtt_connection_failure(self, cause):
         """
 
         logger.error("{}: _on_mqtt_connection_failure called: {}".format(self.name, cause))
-        if self._active_connect_op:
+
+        if isinstance(
+            self._pending_connection_op, pipeline_ops_base.ConnectOperation
+        ) or isinstance(self._pending_connection_op, pipeline_ops_base.ReconnectOperation):
             logger.info("{}: failing connect op".format(self.name))
-            op = self._active_connect_op
-            self._active_connect_op = None
+            op = self._pending_connection_op
+            self._pending_connection_op = None
             op.error = cause
             operation_flow.complete_op(stage=self, op=op)
         else:
             logger.warning("{}: Connection failure was unexpected".format(self.name))
             unhandled_exceptions.exception_caught_in_background_thread(cause)
 
     @pipeline_thread.invoke_on_pipeline_thread_nowait
-    def _on_mqtt_disconnected(self, cause):
+    def _on_mqtt_disconnected(self, cause=None):
         """
         Handler that gets called by the transport when the transport disconnects.
+
+        :param Exception cause: The Exception that caused the disconnection, if any (optional)
         """
         logger.error("{}: _on_mqtt_disconnect called: {}".format(self.name, cause))
 
         # self.on_disconnected() tells other pipeilne stages that we're disconnected.  Do this before
         # we do anything else (in case upper stages have any "are we connected" logic.
         self.on_disconnected()
 
-        # regardless of the cause, we wrap it in a ConnectionDroppedError object because that's
-        # the real problem at this point.
-        if cause:
-            try:
-                six.raise_from(errors.ConnectionDroppedError, cause)
-            except errors.ConnectionDroppedError as e:
-                cause = e
-
-        if self._active_disconnect_op:
+        if isinstance(self._pending_connection_op, pipeline_ops_base.DisconnectOperation):
             logger.info("{}: completing disconnect op".format(self.name))
-            op = self._active_disconnect_op
-            self._active_disconnect_op = None
-            op.error = cause
+            op = self._pending_connection_op
+            self._pending_connection_op = None
+
+            if cause:
+                # Only create a ConnnectionDroppedError if there is a cause,
+                # i.e. unexpected disconnect.
+                try:
+                    six.raise_from(errors.ConnectionDroppedError, cause)
+                except errors.ConnectionDroppedError as e:
```

Comment:
So, this problem:

Native exception chaining was introduced in Python 3, thus isn't compatible with 2. This use of six allows us to use it in 3, while ignoring it in 2. But I wonder if it might be better to use our own implementation of this concept that will be compatible across versions.

i.e. we define our custom exceptions with a `.cause` attribute, and put a space for it in the constructor. Thus we can chain in Python 2.

This will ensure we have consistent tools for investigating errors across versions - it would suck for everything to be harder to investigate in Python 2.

Not suggesting we do it in this PR or anything, this is quite clearly part of a separate PR (or it's own one entirely), but this is something we should take into consideration when putting together our larger error story in the coming weeks.